Kubernetes deployer implementation

frosenberg opened the pull request:
This is a basic Kubernetes deployer implementation. It deploys streams modules as Kubernetes replication controllers and for those sources or sinks that have a `server.port` or `port` parameter (like http) an external service is created (this heuristic may be improved over time). 

A Readme with setup instructions and a simple integration test is provided.

ericbottard commented:
> for those sources or sinks that have a server.port or port parameter (like http)

Note that currently all modules are by default web servers, whether it is central to the module functionality (http) or just for exposing actuators.

Still need to have a look at your code to see how you made the decision to expose this (or not)

ericbottard commented on spring-cloud-dataflow-admin/src/test/java/org/springframework/cloud/dataflow/admin/AdminApplicationTests.java:
```diff
@@ -64,5 +65,15 @@ public void testLatticeProfile() {
+	public void testKubernetesProfile() {
+		SpringApplication app = new SpringApplication(AdminApplication.class);
+		ConfigurableApplicationContext context = app.run(new String[] { "--spring.profiles.active=kubernetes",
+				"--server.port=0" });
```
Can set port to -1 to even not bother creating a webserver (speeding up the test). May apply to other tests

frosenberg replied:
I followed what was there ;). Will change it ...

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
```diff
@@ -0,0 +1,176 @@
+SPI implementation for deploying Spring Cloud Stream modules to [Kubernetes](http://http://kubernetes.io/).
```
Could you please convert this to asciidoc, which we try to use throughout

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
````diff
@@ -0,0 +1,176 @@
+$ ./kube-up.sh
+```
+
+Once the cluster is up, configure you local system to your local DNS server `10.0.0.10` 
````
I don't understand this sentence. There is a typo for sure 'configure *you* local system', but even if it read "your", not sure what to do

frosenberg replied:
Ok, that was bad phrasing. A kubernetes cluster comes with a DNS service. If you add the Kubernetes DNS `10.0.0.10` to your local DNS config (e.g. /etc/resolv.conf on Linux), you can resolve all Kubernetes provisioned services from your development machine (needed later in the Readme).

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
````diff
@@ -0,0 +1,176 @@
+1. Start the Redis server within your Kubernetes cluster:
+
+   ```
+   $ kubectl create -f src/main/kubernetes/redis-controller.yml
````
Not sure `src/main/kubernetes` is the best place for those kind of files. Will need to have a look.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
````diff
@@ -0,0 +1,176 @@
+
+1. Run integration tests (optional)
+
+	 ```
````
Will change once converter to asciidoc anyway, but there seems to be an indentation issue here

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
````diff
@@ -0,0 +1,176 @@
+   ```
+	 $ kubectl get pods -l scsm-label=log
+	 NAME                    READY     STATUS    RESTARTS   AGE
+	 deployment1-log-tacrf   1/1       Running   0          7m 
````
not familiar enough with kubernetes (yet :) ). What is that `-tacrf` suffix?

frosenberg replied:
Kubernetes has the concepts of a replicationController who is responsible for scheduling a set of so-called pods (instance of your service). Those get this kind of weird IDs appended :).

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
```diff
@@ -0,0 +1,176 @@
+
+Internally the Kubernetes Deployer uses the excellent [Kubernetes Client API](https://github.com/fabric8io/kubernetes-client).
+
+If you have a working `kubectl` installation that points to an exsting Kubernetes cluster, the Spring Admin + the Kubernetes deployer will pick up the configuration automatically (in most cases). If not, see the configuration options below.
```
What does that mean? Does it read files from well-known locations?

frosenberg replied:
`kubectl` read config from you ~/.kube/config. This tool is only for managing and working with the cluster. The deployer does not use it.

The Kubernetes Client API, makes some intelligent guesses and in most cases no configuration is needed. Overrides are done via the environment variables. I did not want to replicate all this with configuration properties).

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/README.md:
```diff
@@ -0,0 +1,176 @@
+
+System properties are preferred over environment variables. The following system properties & environment variables can be used for configuration:
+
+* `kubernetes.master` / `KUBERNETES_MASTER`
```
We're working on supporting common properties for all deployers (like setting the amount of RAM for a module, _etc._) but I see this deployer implementation does not support that at all. Is this by design?

Also, about the props below, are those things that are recognized by other tools of the kubernetes ecosystem (_e.g._ `kubectl` installation as mentioned above). If so, we should think about whether or not using the same `kubernetes.` prefix is a good idea

frosenberg replied:
I could not find an example on this. What is supported is the deployment property `module.*.count` which will then launch `count` number of Kubernetes pods that will be managed by the corresponding replication controller.

ericbottard replied:
Sorry, what I meant to say is that we're working on harmonizing the way all deployers specify features of what they deploy (like say, RAM allocated to container) with
1. a global default value
2. a way to override it with deployment properties (`module.*.xxx`)

In any case, your implementation does not seem to even set (even an hard-coded value) for say, the amount of RAM to allocate to the docker container.

frosenberg replied:
Do you have an example?

ericbottard replied:
Have a look at those lines in the [MarathonDeployer](https://github.com/spring-cloud/spring-cloud-dataflow/blob/93bdcb6eb9f058c42e9b4a7e659cf2ea95f9446b/spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-marathon/src/main/java/org/springframework/cloud/dataflow/module/deployer/marathon/MarathonModuleDeployer.java#L97-L105), or those in the [CF deployer](https://github.com/spring-cloud/spring-cloud-dataflow/blob/be5cd0b74dcb93566563c946cad8509e16c56117/spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-cloudfoundry/src/main/java/org/springframework/cloud/dataflow/module/deployer/cloudfoundry/ApplicationModuleDeployer.java#L91-L100)

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/pom.xml:
```diff
@@ -0,0 +1,69 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-stream-module-launcher</artifactId>
```
This dependency should not be needed, conceptually

frosenberg replied:
The `spring-cloud-stream-module-launcher` one?

ericbottard replied:
yes. I haven't actually compiled your code yet, just browsing. But if you need it because you use some API from `spring-cloud-stream-module-launcher`, then this is "wrong" as we want those two to be decoupled.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/pom.xml:
```diff
@@ -0,0 +1,69 @@
+		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-failsafe-plugin</artifactId>
```
Maybe this configuration could eventually be pulled up to the s-c-d level.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+		Integer externalPort = 8080;
+		// we also create a service in case have an '--port' or '--server.port' argument (source or sink)
+		Map<String, String> parameters = request.getDefinition().getParameters();
+		if (parameters.containsKey(PORT_KEY)) {
```
Ideally this would not be "hardcoded" like this. A module may even have multiple ports, etc

frosenberg replied:
How are multiple ports specified? The question here when you look at an entire stream definition, what do you need to expose publicly. Usually, the source in case message need to be retrieved and maybe the sink (in case message need to go out). This is what I tried to replicate but room for improvement.

frosenberg replied:
For Kubernetes, I need to know what ports are "available" to open them. Without metadata about a module that I can query at runtime, I have to make an "educated" guess :)

ericbottard replied:
I see. This is actually not very different from what we do in other deployers, although we use a different approach:
we force (or at least try to leave the value to its default, 8080) the port and ask for a NAT mapping targeting that value

Does kubernetes support raw tcp NAT, or just http proxyfication ?

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerConfiguration.java:
```diff
@@ -0,0 +1,39 @@
+@EnableConfigurationProperties(KubernetesModuleDeployerProperties.class)
+public class KubernetesModuleDeployerConfiguration {
+	
+	@Autowired
```
Can move the (auto)wiring of the `KubernetesModuleDeployerProperties ` object in `KubernetesModuleDeployer` itself, as it is only used there.

frosenberg replied:
+1

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerConfiguration.java:
```diff
@@ -0,0 +1,39 @@
+	
+	@Bean
+	public ModuleDeployer taskModuleDeployer() {
+		return processModuleDeployer();
```
Does kubernetes have support for short-lived processes?

frosenberg replied:
Not sure if it really does. I guess one can build a short lived task as just a Pod.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerConfiguration.java:
```diff
@@ -0,0 +1,39 @@
+	}
+	
+	@Bean
+	public KubernetesClient kubeClient() {
```
Layman's question: is `kube` a known nickname for the library you've used? If not, why not spell it out entirely?

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+	 * Use the default spring image. Override with --kubernetes.moduleLauncherImage 
+	 * as a Spring Admin parameter to use a different one.
+	 */
+	private static final String DEFAULT_IMAGE_NAME = "springcloud/stream-module-launcher";
```
This should not be here, but should be the initialization value in `KubernetesModuleDeployerProperties `

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/resources/rebel.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file should not be in version control

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+
+	@Before 
+	public void setup() {
+		assertNotNull(processModuleDeployer);
```
This can't happen, by design of how `@Autowired` works

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+		String label = "http";
+		
+		ModuleDefinition d = new ModuleDefinition.Builder()
+				.setName("foobar")
```
label and name are actually semantically swapped in your test. Does not matter given that you force the maven coordinates anyway, but what you meant is most certainly the other way around (or there is a bug in s-c-d)

frosenberg replied:
Could be. However, I was confused because if you look at the ModuleDeploymentId, it seems to only have `group` and `label` and no `name`. When you derive a `ModuleDeploymentId` from a `ModuleDefinition`  with `ModuleDeploymentId.fromModuleDefinition(..)` it seems to use the `label` as well and not the `name`.

I think this means it is a bug then.

ericbottard replied:
This is normal.
The label is used to disambiguate two uses of the same (named) module in a stream. Its default value is the module name.

Consider this:
```
http | filter | log
```

there are 3 modules, whose names (and hence labels) are `http`, `filter` and `log`.

In this, however, we need to label one instance of the `filter` module:
```
http | filter | foo: filter | log
```

the 3rd module has `name = filter, label = foo`

In a given stream, the `streamName, moduleLabel` is guaranteed to be unique

frosenberg replied:
Ok I see. I need to figure out how this will work. Need to do some more testing.

ericbottard replied:
As I said above, this is actually irrelevant for your test. I was just pointing it out. Just swap the two values for peace of mind, but this should not change anything

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+		String group = "deployment-test-0";
+		String label = "http";
+		
+		ModuleDefinition d = new ModuleDefinition.Builder()
```
Could you use longer variable names?

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+		assertSame(id, status.getModuleDeploymentId());
+		
+		
+		// wait for "deploying" state. 
```
Why test on a different thread?

frosenberg replied:
I wanted a timeout

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+		
+		
+		// wait for "deploying" state. 
+		final Future<ModuleStatus.State> future = executor.submit(new Callable<ModuleStatus.State>() {
```
Couldn't this test fail if we don't even see the `deploying` state (_i.e._ if it deployed really quickly)?

frosenberg replied:
The whole thing will timeout after 15sec if you don't see it.

ericbottard replied:
yes, but is that a good thing? I mean, a deployer that **does** deploy (very quickly) is a **working** deployer.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+			assertEquals(ModuleStatus.State.deploying, future.get(15000, TimeUnit.MILLISECONDS));
+		} catch (TimeoutException e) {
+		    future.cancel(true);
+		    fail(e.getMessage());
```
This whole thing may be revisited, but if it is not, just `throw e` here. Preserves stacktrace.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+				
+		processModuleDeployer.undeploy(id);
+		
+		// ensure the corresponding Kubernetes entities have been undeployed
```
Doesn't undeploy require the same fire-and-check call as deploy?

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,120 @@
+		
+		// ensure the corresponding Kubernetes entities have been undeployed
+		Map<String, String> labels = createLabelMap(group, label);
+		assertTrue(kubeClient.services().withLabels(labels).list().getItems().isEmpty());
```
This assumes too much. What if there is other stuff running on the cluster this IT targets?

frosenberg replied:
I does not assume anything except finding the services that have a certain `label`, i.e. the only thing that will match is either the services I deployed (and have assigned the label in the deployer) or nothing. 

If the list is not empty, the undeployment did not succeed. The issue could be that I did not wait long enough. I may have to check the status as well as you correctly say.

ericbottard replied:
I stand corrected. The only thing that would match (putting aside a parallel deployment of the same "stream") is the one your test deployed earlier.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+			kubeClient.services().withName(name).delete();
+		} catch (KubernetesClientException e) {
+			logger.error(e.getMessage(), e);
+			// FIXME what do we throw??
```
Any RuntimeException would be fine, for now. Need to throw something though, obviously :)

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+                .withNewMetadata()
+                	.withName(createKubernetesName(id))  // does not allow . in the name
+                	.withLabels(createIdMap(id))
+                	.addToLabels(SPRING_MARKER_KEY, SPRING_MARKER_VALUE)
```
Are all these really necessary when it seems the ones returned from `createIdMap()` are the only ones used to `filter` s-c-d modules?

frosenberg replied:
They are used in the `status()` method. The ones returned by the `createIdMap()` are specific to a module, whereas the markers are not.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+                .endTemplate()
+                .endSpec().build();        
+       
+        return kubeClient.replicationControllers().create(rc);        
```
indent issue

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+        return kubeClient.replicationControllers().create(rc);        
+	}
+
+	private PodSpec createPodSpec(ModuleDeploymentRequest request, Integer port) {
```
`port` can be `int` instead of `Integer` throughout, it seems

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,306 @@
+		} else {
+			container.withImage(properties.getModuleLauncherImage());
+		}		
+		container.withName(CONTAINER_NAME)
```
Where does that name appear? Could it be dynamic and derived from the deployment request?

frosenberg replied:
I think this is doable.

ericbottard commented:
Overall this looks really promising. I shall test it for real, but the main problem is its tight coupling with redis.

We shall work in de-coupling those two by _e.g._ introducing a Spring Cloud Connector implementation for kubernetes.

By the way, another implementation of a deployer you may be interested in (but may not know about) is living in this branch: https://github.com/spring-cloud/spring-cloud-dataflow/tree/93bdcb6eb9f058c42e9b4a7e659cf2ea95f9446b/spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-marathon

frosenberg commented:
@ericbottard thanks for the thorough review. I can address most of the issues I think.

I totally agree that this is too tied to Redis atm. The current thought process was that since Spring Admin is tied to Redis and the dataflow itself needs Redis as transport, we need to pass on the Redis information to the instance that the Kubernetes deployer creates. Ideally, i want pluggable transports. 

The Spring Cloud Connector part is not yet fully clear to me but I have a look at it. Does this mean that the connector is supposed to bring up supporting services on Kubernetes (such as Redis or Kafka)?

ericbottard commented:
It means that the deployer will dynamically discover which transports are available.
It's really a way to advertise some apps (here pods I guess) running on the "cloud" (here kubernetes) as being particular "services", and expose their credentials, so that other consuming apps (here the module running inside the module launcher) can connect to them.

The lattice and marathon deployer implementations are good, simple examples of this abstraction. Even if naive, the cloud connector implementations for these runtimes allow decoupling.

We'll wall through that together, don't worry. I need to have a way more thorough look at your impl. and kubernetes in general before, though

ericbottard commented:
Another thing we'll need to consider and which is often tricky for deployers (we haven't solved it yet for the Marathon deployer, for example) is to support the `module.index` property: the need for each _instance_ of a deployed module (amongst many when `module.count` > 1) to have its own unique index.

This is required to support partitioning.

As an example, consider `time | log` with `module.log.count = 3`, then is there a way in kubernetes to either explicitly assign a unique index to each instance of the `log` module or, even better, does kubernetes do that for you in some way?

frosenberg commented:
The `module.index` is already properly handled. The kubernetes instance here is called a Pod, and they all get their unique id (the `-tacrf` you mention above).

Do you have some partitioning use cases somewhere or at least something where `module.index` is used?

frosenberg commented:
I have looked in the Marathon deployer for the cloud connector example. I don't think I get it. Maybe I need to spend some more time looking at marathon first.

ericbottard commented:
> The `module.index` is already properly handled. The kubernetes instance here is called a Pod, and they all get their unique id (the `-tacrf` you mention above).

I had a feeling this was related indeed. The "problem" is we'd like to have this be an int in the range `[0..${module.count}]`. Moreover, when pod with index "2" dies, we'd like that its replacement also gets "2" as an index. Some kind of registry between free-form indices and numerical ones may be needed...

> Do you have some partitioning use cases somewhere or at least something where module.index is used?

`module.index` is used implicitly when you do partitioning. You'd try this (_e.g._ with the Cloud Foundry or lattice deployers):

```
dataflow:>stream create foo --definition "http --server.port=8080 | log"

dataflow:>stream deploy --name foo --properties "module.log.count=3,module.http.producer.partitionKeyExpression=payload.length()"
```

Then

```
curl -H "Content-Type: text/plain" http://localhost:8080/messages -d hello
curl -H "Content-Type: text/plain" http://localhost:8080/messages -d hellow
curl -H "Content-Type: text/plain" http://localhost:8080/messages -d helloww

```

ericbottard commented on pom.xml:
```diff
@@ -108,4 +108,16 @@
 	</pluginRepositories>
+	
+	
+	<distributionManagement>
```
I think this should not have been commited

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/ContainerFactory.java:
```diff
@@ -0,0 +1,16 @@
+ * 
+ * @author Florian Rosenberg
+ */
+public interface ContainerFactory {
```
I think it would be better if we kept this PR "simple" (it's already quite big and will become more complicated when trying to uncouple redis).

The concept of `ContainerFactory` should be added as a separate PR, later (if ever -- that is another story)

frosenberg replied:
I understand. However, this is just a simple refactoring and most of the code that was related to creating a "container" moved to the DefaultContainerFactory. We can move out the other one (`PreBaked...`) and the related configuration.

ericbottard replied:
I'm not sure we'd want to merge the `PreBaked..` strategy, so discussion about it should be in a separate PR

frosenberg replied:
Ok, the latest push removed this.

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,202 @@
+
+		try {
+			kubernetesClient.replicationControllers().withName(name).delete();
+			kubernetesClient.services().withName(name).delete();
```
A service may not have been created, at least in the way the PR currently works

frosenberg replied:
Yes, but the `kubernetesClient` tends to swallow that. Will add a test case for that to ensure this is the case - and some docs :)

ericbottard replied:
You mean it would not fall in the catch block below (which would be bad)?
Let's make sure this is indeed the case (with a test) or try to come up with a less aggressive approach. From comments above, this needs to change from `withName()` to "withLabels()" approach anyway, which may indeed silently do nothing when there is no match

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,202 @@
+            	.withName(createKubernetesName(id))  // does not allow . in the name
+            	.withLabels(createIdMap(id))
+            	.addToLabels(SPRING_MARKER_KEY, SPRING_MARKER_VALUE)
+        		.addToLabels(SCSM_ARTIFACT_ID, request.getCoordinates().getArtifactId())
```
now that I know a bit more about kubernetes, why are you using labels here? Seems like this should be done via annotations, if at all. What do you think?

frosenberg replied:
I wanted label because I can query them (labels are supposed to be "identifying"). Annotations is something that I cannot really query. 

Labels have the extra advantage that I can also work with them with `kubectl`. If I'm the admin of a large cluster, they come in handy to query all to dataflows that use a module of a certain version and perform a rolling upgrade.

`kubectl get pods -l scsm-label=http scsm-version=1.1.0.SNAPSHOT`

ericbottard replied:
+1, maybe adding a comment behind the rationale

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,202 @@
+            	.addToLabels(SPRING_MARKER_KEY, SPRING_MARKER_VALUE)
+        		.addToLabels(SCSM_ARTIFACT_ID, request.getCoordinates().getArtifactId())
+        		.addToLabels(SCSM_GROUP_ID, request.getCoordinates().getGroupId())
+        		.addToLabels(SCSM_VERSION, request.getCoordinates().getVersion())
```
If we keep this, you might as well add `classifier`

frosenberg replied:
Thought so but since we always run `exec` it did not matter. But I see your point.

ericbottard replied:
**WE** always use `exec`. Other developers may choose to use anything for their qualifier (if at all). I think this should be added (with `null` guard) for completness

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,202 @@
+        ReplicationController rc = new ReplicationControllerBuilder()
+            .withNewMetadata()
+            	.withName(createKubernetesName(id))  // does not allow . in the name
+            	.withLabels(createIdMap(id))
```
As far as I understand, this labelling is not necessary, at least for the good behavior of the deployer (later retrieval of the rc is done by name only (at `undeploy()`)

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java:
```diff
@@ -0,0 +1,202 @@
+		kubernetesClient.services().inNamespace(kubernetesClient.getNamespace()).createNew()
+			.withNewMetadata()
+              .withName(KubernetesUtils.createKubernetesName(id)) // does not allow . in the name
+              .withLabels(createIdMap(id))
```
ditto as above, this labelling is not necessary, is it?

frosenberg replied:
Bug in my code. it should only be label based not name based. The issue is that the name of a `service` in Kubernetes falls under certain restrictions (see my other issue) so a name needs to be deduced that may not reassemble the `group` and `name`/`label`.

ericbottard replied:
I see your point. While you're at it, you may want to change the name of the `createIdMap()` method / add clarifying javadoc

frosenberg commented:
@ericbottard i'll work on the changes later this week or early next to finish this up. I still need to look into the cloud-connector thing. I cannot find very good samples tbh.

frosenberg commented:
I just read up on the cloud connector approach. If I understand it correctly, it lets me discover cloud services (mostly on PaaSes) that can be dynamically provisioned for you. The cloud connector approach enables you do connect to those by returning the connection info. This assumes that such services are available. On Bluemix or Heroku, they have a fixed set of services one can provision and connect to.

On Kubernetes, that is not the case (services on Kubernetes have a different semantics than services in a PaaS). Since it is a cluster member, the operator can choose what to deploy. There is no such thing as a service catalog where I can go, pick a service and provision it. So, if the operator chooses to deploy a Kafka or a Redis cluster, it might ultimately be exposed as a Kubernetes services (which acts as a loadbalancer and has a DNS name). 

If I write a Cloud connector for Kubernetes, I'm not sure how this maps and what set of "cloud services" in the PaaS sense would be exposed in the implementation. I may be missing a bit here and there :).

markfisher commented:
The cloud connector for Lattice might be the most relevant example. Since there is no first class distinction for "service" (as opposed to "app"), the services are discovered by convention, e.g. starting with an expected prefix. Have a look at the code in this package for an overview: https://github.com/spring-cloud/spring-cloud-lattice/tree/master/spring-cloud-lattice-connector/src/main/java/org/springframework/cloud/lattice/connector

ericbottard commented:
I'll try to submit a PR against your branch with at least an embryo of the cloud connector stuff. For the moment, this shall be created as another maven project in the deployers directory (similar to what is done in the Marathon branch) so that it builds easily.

ericbottard commented:
@markfisher @mbogoevici We really need to decide what to do wrt to the `module.index` thing. There is no easy solution in the Marathong/Kubernetes world. One could decide to create N templates, each with their own `index` property explicitly set, but this kind of defeats the native support for "instances" of those platforms.
Maybe we should resort to some generic support for a central service that maps "native" ids (like the pod suffix here in kubernetes) to an int index, but this brings its share of problems, obviously...
My point being that it will be hard to stall merging branches until we have proper support. Maybe those two deployers should refuse to deploy when there are partitioning related props, for now.

mbogoevici commented:
@ericbottard my initial reluctance was caused by the feeling that this was local to the Marathon deployer (and based on the notion that we could solve it through our own scheduler). However, if there is a class of systems where this applies, then it is more visible, and it becomes quickly apparent that we would need to address this in a generic fashion. Maybe we can reconsider our use of instance indexes altogether - but I would agree that it would be beyond the scope of this PR.

markpollack commented:
Maybe instead of passing it in, it should be looked up by the module itself in a first come first serve manner.  Not sure yet what would back that...

ericbottard commented:
@markpollack  yes, this is what I was talking about. Comes with a lot of problems though (how to be notified of failed nodes, etc). Let's create a document to discuss this

frosenberg commented:
@ericbottard: I added all the memory and cpu stuff and removed the PreBackedImage stuff as discussed. You can now do the following:

```
stream create --name deployment2 --definition "http --server.port=9999 | log"
stream deploy --name deployment2 --properties "module.http.kubernetes.cpu=1,module.http.kubernetes.memory=1000Mi"
```




I think what is left is cloud connector and `module.index` stuff?

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerProperties.java:
```diff
@@ -0,0 +1,153 @@
+	 * Delay in seconds when the Kubernetes liveness check of the stream module container 
+	 * should start checking its health status.
+	 * 
+	 * See here for more information: 
```
All those properties will get documentation metadata information generated thanks to http://docs.spring.io/spring-boot/docs/1.3.0.RC1/reference/htmlsingle/#configuration-metadata

The javadoc should thus be kept to a "minimum", avoiding special tags, etc.
Additional information can be added as a regular (non javadoc) comment

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerProperties.java:
```diff
@@ -0,0 +1,153 @@
+	 * Timeout in seconds for the Kubernetes liveness check of the stream module container. 
+	 * If the health check takes longer than this value to return it is assumed as 'unavailable'.
+	 * 
+	 * See here for more information: 
```
same here

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerProperties.java:
```diff
@@ -0,0 +1,153 @@
+	 * Timeout in seconds that the stream module container has to respond its
+	 * health status during the readiness check. 
+	 * 
+	 * See here for more information: 
```
same here

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerProperties.java:
```diff
@@ -0,0 +1,153 @@
+	 * Delay in seconds when the readiness check of the stream module container 
+	 * should start checking if the module is fully up and running.
+	 * 
+	 * See here for more information: 
```
same here

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerProperties.java:
```diff
@@ -0,0 +1,153 @@
+	private int readinessProbeTimeout = 2;
+
+	/**
+	 * Memory to allocate for a Pod
```
Missing dot at the end

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerProperties.java:
```diff
@@ -0,0 +1,153 @@
+	
+
+	public String getModuleLauncherImage() {
+		return moduleLauncherImage == null ? DEFAULT_IMAGE_NAME : moduleLauncherImage;
```
Just initialize the field with `DEFAULT_IMAGE_NAME `

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/kubernetes/redis-service.yml:
```diff
@@ -0,0 +1,9 @@
+  ports:
+    - port: 6379
+  selector:
+    name: redis
```
missing newline

ericbottard commented on spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/test/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployerIT.java:
```diff
@@ -0,0 +1,110 @@
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = KubernetesModuleDeployerConfiguration.class)
+public class KubernetesModuleDeployerIT {
```
We'll want to make this conditional to the fact that there is a kubernestes cluster running where we expect it to.
See for example a JUnit rule to only run ITs thats require redis when redis is available: https://github.com/spring-cloud/spring-cloud-stream/blob/7023e161967da795ab5f94fde6b4a5964d8c1aa2/spring-cloud-stream-test-support-internal/src/main/java/org/springframework/cloud/stream/test/junit/redis/RedisTestSupport.java

I think this can go in a different PR though

ericbottard commented:
@frosenberg Yes! This is definitely taken shape, again thanks for all your efforts!

This is the list of things I think still need to be taken care of, not necessarily in this PR:
* Decouple from redis using spring cloud connector
* Don't run the IT unless there is a kubernetes cluster available
* Fail deployment if trying to use partitioning (for now)
   * support `module.index` in some way
* I think we'll also want to using the Spring Boot mechanisms to configure the client, instead of relying on its native way. That means passing in a `Config` object which was created out of props from a/the `@ConfigurationProperties` class. This to harmonize configuration between all deployers and be consistent in the whole project.

I'll try to work on item n1 tomorrow. In the meantime, if you haven't done so already, could you rebase this PR on top of the latest master?

ericbottard commented:
Hmm, I see you have squashed your commits, but I'm not sure I got them all at the time I started working on the spring cloud connector.
Doesn't matter, I'll open a new PR and we can continue the discussion from there.

ericbottard commented:
Closing this, with further discussion expected to happen in https://github.com/spring-cloud/spring-cloud-dataflow/pull/132

Thanks again for the great work @frosenberg

frosenberg commented:
Thank @ericbottard. I think you should rebase on my latest squashed commit if possible. I think I made some changes that are important.
